Update CubicSupercellTransformation and merge PerturbSitesTransformation into PerturbStructureTransformation

rees-c opened the pull request:
## Summary
- Added a boolean argument to CubicSupercellTransformation that can constrain the transformation matrix to be diagonal
- Made a bug fix in CubicSupercellTransformation regarding how the smallest dimension is found
- Merged PerturbSitesTransformation into PerturbStructureTransformation (as discussed in #1536)
  - Removed PerturbSitesTransformation and its unit test from advanced_transformations
  - Added an option to PerturbStructureTransformation and Structure.perturb that allows for the displacements to be randomly distributed between two values.
  - Added to unit tests for PerturbStructureTransformation and Structure.perturb() to reflect the new option
  - The rest of the advanced features of the old PerturbSitesTransformation will be implemented into my CSLD atomate workflow by calling the standard transformation multiple times
  - Changed the args of PerturbStructureTransformation to match those of Structure.perturb (i.e. changed `amplitude` to `distance`). *****NOTE THIS IS POTENTIALLY A BREAKING CHANGE FOR ANY ATOMATE WORKFLOW USES `PerturbStructureTransformation`.**

utf commented on pymatgen/core/structure.py:
```diff
@@ -3178,21 +3178,31 @@ def rotate_sites(self, indices=None, theta=0, axis=None, anchor=None,
             vector = np.random.randn(3)
             vnorm = np.linalg.norm(vector)
-            return vector / vnorm * distance if vnorm != 0 else get_rand_vec()
+            if isinstance(min_random_distance, float) or \
```
You can write these two lines as just:

```python
 if isinstance(min_random_distance, (float, int)):
```

rees-c replied:
Oh nice, didn't know you could do that

utf commented on pymatgen/transformations/standard_transformations.py:
```diff
@@ -680,15 +680,20 @@ class PerturbStructureTransformation(AbstractTransformation):
     Args:
         amplitude (float): Amplitude of perturbation in angstroms. All sites
             will be perturbed by exactly that amplitude in a random direction.
+        min_random_distance (None, int, or float): If None, perturb
```
For the transformation, can you update this to be `min_random_amplitude` to match the other argument names.

Also edit the rest of the docstring accordingly.

utf commented on pymatgen/transformations/standard_transformations.py:
```diff
@@ -680,15 +680,20 @@ class PerturbStructureTransformation(AbstractTransformation):
     def apply_transformation(self, structure):
         s = structure.copy()
-        s.perturb(self.amplitude)
+        s.perturb(self.amplitude, self.min_random_distance)
```
to avoid any errors arising from new variables being added to `Structure.perturb`, can you make this line:

```python
s.perturb(self.amplitude, min_random_distance=self.min_random_ampliude)
```

utf commented:
You should mention that your way of merging the two classes was to:
- Remove the advanced transformation.
- Add an option to `PerturbStructureTransformation` and `Structure.perturb` that allows for the displacements to be randomly distributed between two values.

The rest of the advanced features of the transformation will be implemented in your own code by calling the standard transformation multiple times.

utf commented on pymatgen/core/structure.py:
```diff
@@ -3178,21 +3178,30 @@ def rotate_sites(self, indices=None, theta=0, axis=None, anchor=None,
             vector = np.random.randn(3)
             vnorm = np.linalg.norm(vector)
-            return vector / vnorm * distance if vnorm != 0 else get_rand_vec()
+            if isinstance(min_random_distance, (float, int)):
```
In fact this whole thing could be rewritten much more nicely as:

````python
 if isinstance(min_random_distance, (float, int)):
    distance = np.random.uniform(min_random_distance, distance)

return vector / vnorm * distance if vnorm != 0 else get_rand_vec()
```

rees-c replied:
So for some reason when I tried that code, it treats `distance` inside `np.random.uniform(min_random_distance, distance)` as an unresolved variable local to get_rand_vec(). Would this be an ok solution:

```python
dist = distance
if isinstance(min_random_distance, (float, int)):
     dist = np.random.uniform(min_random_distance, dist)
return vector / vnorm * dist if vnorm != 0 else get_rand_vec()
```

mkhorton commented:
Thanks for working on this, it's looking much better! Quick note that since we're modifying the `.perturb()` method there will also need to be a test in `pymatgen.core` for the method also, otherwise after @utf's comments are resolved I think this looks good.

mkhorton commented:
Perhaps one way to improve legibility would be to change the arg name on `perturb` to `min_amplitude` and say "if None, all displacements will be equal amplitude" (just a small note because we're changing from using the word "amplitude" to the word "random_distance" to describe essentially the same thing).

utf commented:
@mkhorton currently the main perturb argument is called “distance”, should we also change this to amplitude? 

Personally, I think distance is a more obvious naming choice. However, if we were to use this throughout it would be a breaking change as amplitude is a named argument in the transformation class.

mkhorton commented:
Ah, apologies -- I was looking at the StandardTransformation args, not the perturb args. I'd want to rename that ideally to be consistent with the perturb argument not sure if it will break backwards compatibility somewhere though. (I have no opinion on "amplitude" vs "distance", as long as it's consistent)

rees-c commented:
Hey @mkhorton ! I implemented alex's comments, added to the Structure.perturb unit test, changed the doc string wording for `perturb` and `PerturbStructureTransformation` to "if None, all displacements will be equal amplitude", and refactored `amplitude` and `min_amplitude` in `PerturbStructureTransformation` to `distance` and `min_distance` (to match the arg names in `Structure.perturb`). Does this look ok? I can undo the refactoring of `amplitude` and `min_amplitude` if we think that might break something

mkhorton commented:
As long as tests pass I’m ok with this change, it’s not a big difference if
we need to update something in atomate etc. Thanks again!

On Thu, Aug 1, 2019 at 12:24, rees-c <notifications@github.com> wrote:

> Hey @mkhorton <https://github.com/mkhorton> ! I implemented alex's
> comments, added to the Structure.perturb unit test, changed the doc string
> wording for perturb and PerturbStructureTransformation to "if None, all
> displacements will be equal amplitude", and refactored amplitude and
> min_amplitude in PerturbStructureTransformation to distance and
> min_distance (to match the arg names in Structure.perturb). Does this
> look ok? I can undo the refactoring of amplitude and min_amplitude if we
> think that might break something
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/materialsproject/pymatgen/pull/1544?email_source=notifications&email_token=AAWWWRHNLV42QG2YFNABH4DQCM2AHA5CNFSM4IITALOKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3LUHCA#issuecomment-517424008>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAWWWRHM3YUPG5TCDLM2YCTQCM2AHANCNFSM4IITALOA>
> .
>

rees-c commented:
Thanks for your help @mkhorton!

rees-c commented:
Hi @mkhorton ! The checks have passed so I think we're ready to merge :)
